File: 2020/15_RambunctiousRecitation/memory.lua
-- ======================================================================
-- Rambunctious Recitation
--   Advent of Code 2020 Day 15 -- Eric Wastl -- https://adventofcode.com
--
-- lua implementation by Dr. Dean Earl Wright III
-- ======================================================================

-- ======================================================================
--                         m e m o r y . l u a
-- ======================================================================
-- Memory for the Advent of Code 2020 Day 15 puzzle

-- ----------------------------------------------------------------------
--                                                                  local
-- ----------------------------------------------------------------------
local Memory = { part2=false, text='', age=nil, turn=0, numbers={} }

-- ----------------------------------------------------------------------
--                                                              constants
-- ----------------------------------------------------------------------

-- ======================================================================
--                                                                 Memory
-- ======================================================================


function Memory:Memory (o)
  -- Object for Rambunctious Recitation


  -- 1. Set the initial values
  o = o or {}
  o.part2 = o.part2 or false
  o.text = o.text or {}
  o.age = nil
  o.turn = 0
  o.numbers = {}

  -- 2. Create the object metatable
  self.__index = self
  setmetatable(o, self)

  -- 3. Process text (if any)
  if o.text ~= nil and #o.text > 0 then
     o:_process_text(o.text)
  end

  return o
end

function Memory:_process_text(text)
  -- Assign values from text

  -- 0. Precondition axioms
  assert(text ~= nil and #text > 0)
  
  -- 1. Loop for all of the numbers given
  for number in text:gmatch("[0-9]+") do
    
    -- 2. Remember the number
    self:add(tonumber(number))
  end
  
end

function Memory:add_last_spoken()
  -- Add the last number spoken
  self:add(self.age)
end

function Memory:add(number)
  -- Add a number to the memory
  
  -- 1. Increment the turn number
  self.turn = self.turn + 1
  
  -- 2. If the number is new, the age is 0
  if nil == self.numbers[number] then
    self.age = 0
    
  -- 3. Else the age is calculated from when last seen
  else
    self.age = self.turn - self.numbers[number]
  end
  
  -- 4. Save when the number was last seen
  self.numbers[number] = self.turn
end

    
-- ----------------------------------------------------------------------
--                                                  module initialization
-- ----------------------------------------------------------------------
return Memory

-- ======================================================================
-- end                       m e m o r y . l u a                      end
-- ======================================================================

File: 2020/15_RambunctiousRecitation/memory.ts
// ======================================================================
// Rambunctious Recitation
//   Advent of Code 2020 Day 15 -- Eric Wastl -- https://adventofcode.com
//
// TypeScript implementation by Dr. Dean Earl Wright III
// ======================================================================

// ======================================================================
//                           m e m o r y . t s
//
// Memory for the Advent of Code 2020 Day 15 problem
// ======================================================================

// ----------------------------------------------------------------------
//                                                                  types
// ----------------------------------------------------------------------
type Spoken = Record<number, number>;

// ----------------------------------------------------------------------
//                                                              constants
// ----------------------------------------------------------------------
const RE_NUM = RegExp('[0-9]+', 'g');

// ======================================================================
//                                                                 Memory
// ======================================================================

export class Memory {
  // Object for Rambunctious Recitation
  text: string;

  part2: boolean;

  age: number;

  turn: number;

  numbers: Spoken;

  constructor(text: string, part2 = false) {
    // Create a Memory object

    // 1. Set the initial values
    this.text = text === undefined ? '' : text;
    this.part2 = part2 === undefined ? false : part2;
    this.age = -1;
    this.turn = 0;
    this.numbers = {};

    // 2. Process text (if any)
    if (this.text.length !== 0) {
      const matches = Array.from(this.text.matchAll(RE_NUM), (m) => m[0]);
      for (let indx = 0; indx < matches.length; indx += 1) {
        this.add(parseInt(matches[indx], 10));
      }
    }
  }

  add_last_spoken() {
    // Add the last number spoken
    this.add(this.age);
  }

  add(num: number) {
    // Add a number to the memory

    // 1. Increment the turn number
    this.turn += 1;

    // 2. If the number is new, the age is 0
    if (undefined === this.numbers[num]) {
      this.age = 0;
    } else {
      // 3. Else the age is calculated from when last seen
      this.age = this.turn - this.numbers[num];
    }

    // 4. Save when the number was last seen
    this.numbers[num] = this.turn;
  }
}

// ======================================================================
// end                        m e m o r y . t s                       end
// ======================================================================
